### PR TITLE
Roots of polynomials with Complex Coefficients

### PR DESCRIPTION
**Pull request by bCosgrovePayne**
Added function to FindRoots class for polynomials with complex coefficients, based on existing .Roots() function in Polynomial class. 

I hope I did this right!

### CODE REVIEW

File: src/Numerics/FindRoots.cs [hunk 1]
```diff
@@ -118,6 +122,39 @@ public static Complex[] Polynomial(double[] coefficients)
             return new Polynomial(coefficients).Roots();
         }
 
+        /// <summary>
+        /// Find all roots of a polynomial with complex coefficients by calculating the characteristic polynomial of the companion matrix
+        /// </summary>
+        /// <param name="coefficients">The coefficients of the polynomial in ascending order, e.g. new Complex[] {new (5, 0), new (0, 2) new (2,3)} = "5 + 2i x^1 + (2 + 3i) x^2"</param>
```
Comment (jkalias) [inline src/Numerics/FindRoots.cs]: `new (0, 2) new (2,3)` -> `new (0, 2), new (2,3)` (a comma is missing)

File: src/Numerics/FindRoots.cs [hunk 2]
```diff
@@ -118,6 +122,39 @@ public static Complex[] Polynomial(double[] coefficients)
...
+        /// <returns>The roots of the polynomial</returns>
+
+        public static Complex[] Roots(Complex[] coefficients)
+        {
+            int n = coefficients.Length;
+            if (n < 2)
+            {
+                return null;
```
Comment (jkalias) [inline src/Numerics/FindRoots.cs]: Wouldn't it be better to return an empty array/list instead of null?

**Review by jkalias — CHANGES_REQUESTED** [review]
Can you please add some unit tests covering various tests cases?

**Comment by bCosgrovePayne** [comment]
> Can you please add some unit tests covering various tests cases?

I'll get on that today and try my best! Sorry for any inconvenience, this is my first real pull request.

**Comment by jkalias** [comment]
> > Can you please add some unit tests covering various tests cases?
> 
> I'll get on that today and try my best! Sorry for any inconvenience, this is my first real pull request.

No worries, we all start somewhere :)

File: src/Numerics/FindRoots.cs [hunk 3]
```diff
@@ -118,6 +122,39 @@ public static Complex[] Polynomial(double[] coefficients)
...
+            for (int i = n - 1; i >= 0; i--)
+            {
+                p[i] = -coefficients[i] / aN;
+            }
+
+            DenseMatrix A0 = DenseMatrix.CreateDiagonal(n - 1, n - 1, 1.0);
+            DenseMatrix A = new DenseMatrix(n);
+
+            A.SetSubMatrix(1, 0, A0);
+            A.SetRow(0, p.Reverse().ToArray());
```
Comment (jkalias) [inline src/Numerics/FindRoots.cs:155]: Shouldn't this be the last column, based on this wiki article?
https://en.wikipedia.org/wiki/Companion_matrix

Unless you decide to set the submatrix from the second column and first row. Or maybe I am confused :)

File: src/Numerics.Tests/RootFindingTests/FindRootsTest.cs [hunk 4]
```diff
@@ -210,5 +211,44 @@ public void QuadraticExpected(double c, double b, double a, double x1R, double x
...
+
+        [Test]
+        public void RootsTest()
+        {           
+            var c1 = new Complex[] { new Complex(1, 1), new Complex (1,1), new Complex (1,1) };
+
+            var r1 = FindRoots.Roots( new Complex[] { c1[0] });
+            Assert.AreEqual(Array.Empty<Complex>(), r1);
+
+            var r2 = FindRoots.Roots(new Complex[] { c1[0], c1[1] });
```
Comment (jkalias) [inline src/Numerics.Tests/RootFindingTests/FindRootsTest.cs]: This is supposed to be a first order polynomial, right? So we should have a single root, shouldn't we?

File: src/Numerics.Tests/RootFindingTests/FindRootsTest.cs [hunk 5]
```diff
@@ -210,5 +211,44 @@ public void QuadraticExpected(double c, double b, double a, double x1R, double x
             AssertComplexEqual(Complex.Zero, c + b*x1 + a*x1*x1, 1e-14);
             AssertComplexEqual(Complex.Zero, c + b*x2 + a*x2*x2, 1e-14);
         }
+
+        [Test]
+        public void RootsTest()
```
Comment (jkalias) [inline src/Numerics.Tests/RootFindingTests/FindRootsTest.cs]: Please separate the test cases in separate test functions. This way it's much easier to follow what the edge cases are and what the normal/expected behavior is.

**Review by jkalias — CHANGES_REQUESTED** [review]

File: src/Numerics/FindRoots.cs (hunk 3 above)
Comment (bCosgrovePayne) [reply src/Numerics/FindRoots.cs:155]: So to be entirely honest, I don't know the theory. This is based entirely on the existing .Roots() property in the Polynomial class, which uses Evd to approximate the roots when the polynomial is higher than degree 3, if I'm not mistaken. I monkeyed around with the complex version of the DenseMatrix class and found out everything still works. 

I'll triple check that this is producing accurate roots, but it seemed to pass the unit test w/o issue.

File: src/Numerics.Tests/RootFindingTests/FindRootsTest.cs (hunk 4 above)
Comment (bCosgrovePayne) [reply src/Numerics.Tests/RootFindingTests/FindRootsTest.cs]: So all I know is that when the Polynomial.Roots() property is called and it uses Evd to approximate, it doesn't accept first order polynomials. I wanted to keep everything as close to existing code as possible, but I can definitely add a case for first order polynomials that bypasses the Evd approximation. I could also include quadratic and cubic cases if that's desired.

File: src/Numerics.Tests/RootFindingTests/FindRootsTest.cs (hunk 5 above)
Comment (bCosgrovePayne) [reply src/Numerics.Tests/RootFindingTests/FindRootsTest.cs]: Ok! I'll look around the existing test code more for good examples of what to emulate. I have an idea of what you mean, but it doesn't hurt to keep things as similar to the existing library as possible.

File: src/Numerics.Tests/RootFindingTests/FindRootsTest.cs [hunk 6]
```diff
@@ -210,5 +211,113 @@ public void QuadraticExpected(double c, double b, double a, double x1R, double x
             AssertComplexEqual(Complex.Zero, c + b*x1 + a*x1*x1, 1e-14);
             AssertComplexEqual(Complex.Zero, c + b*x2 + a*x2*x2, 1e-14);
         }
+
+        [Test]
+        public void RootsTestEmpty()
```
Comment (jkalias) [inline src/Numerics.Tests/RootFindingTests/FindRootsTest.cs]: I would keep the naming convention as it currently is, i.e. no 'Test' in the function name (even though I don't agree with this convention)

File: src/Numerics/FindRoots.cs [hunk 7]
```diff
@@ -118,6 +122,48 @@ public static Complex[] Polynomial(double[] coefficients)
...
+        public static Complex[] Roots(Complex[] coefficients)
+        {
+            int n = coefficients.Length - 1;
+            if (n < 1)
+            {
+                return Array.Empty<Complex>();
+            }
+            if (n == 1)
+            {
+                //coefficients[0] = (c + di)
```
Comment (jkalias) [inline src/Numerics/FindRoots.cs]: These comments are not really needed

File: src/Numerics.Tests/RootFindingTests/FindRootsTest.cs [hunk 8]
```diff
@@ -210,5 +211,113 @@ public void QuadraticExpected(double c, double b, double a, double x1R, double x
...
+            }
+        }
+
+        [Test]
+        public void RootstestCubic()
+        {
+            //Expected root values calculated with cubic root forumla using System.Math functions
+            var coefficients = new Complex[] { new Complex(-2, 3), new Complex(1, 0), new Complex(0, 1), new Complex(2, 1) };
+            var expectedRoots = new Complex[] { new Complex(0.828084722641885, -0.6508989159005254), new Complex(-0.9759670820130891, -0.900661591057162), new Complex(-0.052117640628795536, 1.1515605069576873) };
+
```
Comment (jkalias) [inline src/Numerics.Tests/RootFindingTests/FindRootsTest.cs:280]: image
Wolfram Alpha gives different roots, did I do something wrong here?

Comment (jkalias) [reply src/Numerics.Tests/RootFindingTests/FindRootsTest.cs:280]: https://www.wolframalpha.com/input?i=polynomial+roots&assumption=%7B%22C%22%2C+%22polynomial+roots%22%7D+-%3E+%7B%22Calculator%22%2C+%22dflt%22%7D&assumption=%7B%22F%22%2C+%22PolynomialRootCalculator%22%2C+%22polynomial%22%7D+-%3E%22%282%2Bi%29%2B%280%2Bi%29x%2B%281%2B0%29x%5E2%2B%28-2%2B3i%29x%5E3%22
here the link

File: src/Numerics.Tests/RootFindingTests/FindRootsTest.cs [hunk 9]
```diff
@@ -210,5 +211,113 @@ public void QuadraticExpected(double c, double b, double a, double x1R, double x
...
+                AssertComplexEqual(expectedRoots[i], roots[i], 1e-10);
+            }
+        }
+
+        [Test]
+        public void RootsTestQuartic()
+        {
+            //coefficients were derivied from the roots to create a contrived quartic to check
+            var coefficients = new Complex[] { new Complex(-590, 870), new Complex(126, -53), new Complex(-33, -71), new Complex(-9, 22), new Complex(2, 4) };
+            var expectedRoots = new Complex[] { new Complex(-4.5, -1), new Complex(3, -2), new Complex(-4, -3), new Complex(2, 2) };
```
Comment (jkalias) [inline src/Numerics.Tests/RootFindingTests/FindRootsTest.cs:304]: Same here
image

https://www.wolframalpha.com/input?i=polynomial+roots&assumption=%7B%22C%22%2C+%22polynomial+roots%22%7D+-%3E+%7B%22Calculator%22%2C+%22dflt%22%7D&assumption=%7B%22F%22%2C+%22PolynomialRootCalculator%22%2C+%22polynomial%22%7D+-%3E%22%282%2B4i%29%2B%28-9%2B22i%29x%2B%28-33-71i%29x%5E2%2B%28126-53i%29x%5E3%2B%28-590%2B870i%29x%5E4%22

**Review by jkalias — CHANGES_REQUESTED** [review]

File: src/Numerics.Tests/RootFindingTests/FindRootsTest.cs (hunk 8 above)
Comment (bCosgrovePayne) [reply src/Numerics.Tests/RootFindingTests/FindRootsTest.cs:280]: Yes! I'm following the convention in the rest of the FindRoots class, where the coefficients are in ascending order. Basically, in your example, they're reversed. I'll reply to your other examples with the corresponding links 

https://www.wolframalpha.com/input?i=polynomial+roots&assumption=%7B%22C%22%2C+%22polynomial+roots%22%7D+-%3E+%7B%22Calculator%22%2C+%22dflt%22%7D&assumption=%7B%22F%22%2C+%22PolynomialRootCalculator%22%2C+%22polynomial%22%7D+-%3E%22%28-2%2B3i%29%2B%281%2B0%29x%2B%280%2Bi%29x%5E2%2B%282%2Bi%29x%5E3%22

File: src/Numerics.Tests/RootFindingTests/FindRootsTest.cs (hunk 9 above)
Comment (bCosgrovePayne) [reply src/Numerics.Tests/RootFindingTests/FindRootsTest.cs:304]: https://www.wolframalpha.com/input?i=polynomial+roots&assumption=%7B%22C%22%2C+%22polynomial+roots%22%7D+-%3E+%7B%22Calculator%22%2C+%22dflt%22%7D&assumption=%7B%22F%22%2C+%22PolynomialRootCalculator%22%2C+%22polynomial%22%7D+-%3E%22%28-590%2B870i%29%2B%28126-53i%29x%2B%28-33-71i%29x%5E2%2B%28-9%2B22i%29x%5E3%2B%282%2B4i%29x%5E4%22

File: src/Numerics.Tests/RootFindingTests/FindRootsTest.cs (hunk 8 above)
Comment (jkalias) [reply src/Numerics.Tests/RootFindingTests/FindRootsTest.cs:280]: Oh sorry, my bad

**Review by jkalias — APPROVED** [review]

File: src/Numerics.Tests/RootFindingTests/FindRootsTest.cs (hunk 8 above)
Comment (bCosgrovePayne) [reply src/Numerics.Tests/RootFindingTests/FindRootsTest.cs:280]: No worries, thanks for reviewing and your patience!
